File: clj/009.clj
;; Project Euler -- Problem 9

;; A Pythagorean triplet is a set of three natural numbers, a < b < c,
;; for which, a2 + b2 = c2

;; For example, 32 + 42 = 9 + 16 = 25 = 52.

;; There exists exactly one Pythagorean triplet for which a + b + c =
;; 1000.  Find the product abc.

(defn py-triple?
  [t]
  (letfn [(sq [x] (* x x))]
    (let [[ a b c] t]
      (= (sq c) (+ (sq a) (sq b))))))

(println (let [r (range 1 500)]
           (apply * (first (filter #(and (= 1000 (apply + %)) (py-triple? %))
                                   (for [x r y r z r :when (< x y z)] [x y z]))))))

File: pl/009.pl
#!/usr/bin/perl

use 5.10.0;

# Project Euler -- Problem 9

# A Pythagorean triplet is a set of three natural numbers, a < b < c,
# for which, a2 + b2 = c2

# For example, 32 + 42 = 9 + 16 = 25 = 52.

# There exists exactly one Pythagorean triplet for which a + b + c =
# 1000.  Find the product abc.

sub square { $_[0] * $_[0] }
sub is_py_triple {
    my ($a, $b, $c) = @_;
    square($a) + square($b) == square($c)
}

for my $a (1..500) {
    for my $b (1..500) {
        for my $c (1..500) {
            if ($a + $b + $c == 1000 and is_py_triple($a, $b, $c)) {
                say "$a * $b * $c = ", $a * $b * $c;
            }
        }
    }
}
